File: #DFS/dfs.cpp
#include <bits/stdc++.h>
using namespace std;

void DFS(const vector<vector<int>> &graph, unordered_map<int, bool> &visited, int source){
    cout << source << endl;
    visited[source] = true;
    for(auto neighbor : graph[source])
        if(!visited[neighbor]) 
            DFS(graph, visited, neighbor);
}
int main()
{
    int V, E;
    cout << "Input number of vertices: " << endl;
    cin >> V;
    cout << "Input number of edges: " << endl;
    cin >> E;
    vector<vector<int>> graph(V);
    unordered_map<int, bool> visited;
    for(int i = 0; i < E; i++){
        int u, v;
        cout << "Input the " << i+1 << ". edge (two vertices):\n";
        cin >> u;
        cin >> v;
        graph[u].push_back(v);
        graph[v].push_back(u);
    }
    int source;
    cout << "Start DFS from node: ";
    cin >> source;
    DFS(graph, visited, source);
}

File: #DFS_py/dfs.py
from collections import deque

def dfs(graph, visited, source):
    print(source)
    visited[source] = True 
    for neighbor in graph[source]:
        if not visited[neighbor]:
            dfs(graph, visited, neighbor)

def bfs(graph, visited, distance, source):
    q = deque()
    q.append(source)
    distance[source] = 0
    visited[source] = True
    while q:
        u = q.popleft()
        print(u)
        for neighbor in graph[u]:
            if not visited[neighbor]:
                visited[neighbor] = True
                q.append(neighbor)
                distance[neighbor] = distance[u] + 1


def main():
    V = int(input("Number of vertices: "))
    E = int(input("Number of edges: "))
    graph = [[] for _ in range(V)]
    visited = [False] * V
    for i in range(E):
        u, v = map(int, input(f"Input {i}th edge: ").split())
        graph[u].append(v)
        graph[v].append(u)
    
    for v in range(V):
        if not visited[v]:
            dfs(graph, visited, v)
    
    visited = [False] * V
    distances = [-1] * V
    bfs(graph, visited, distances, 0)

    for i, d in enumerate(distances):
        print(f"{i}: {d}")

main()
